### ISSUE TITLE
searchParams on server component/page in nextjs

### ISSUE THREAD

**Issue by jimjamdev** [asmyshlyaev177/state-in-url#40]
**Is your feature request related to a problem? Please describe.**
I'm trying to get the searchParams on a nextjs server page, and pass it to a fetch function. But searchParams are never updated

**Describe the solution you'd like**
I'd like to get the latest searchParams within a server component

**Additional context**

```
import { PageHeading } from '@/components/layouts/page/page-heading';
import { Container } from '@/components/layouts/page/container';

import { SearchParams } from '@/actions/global.schemas';
import { queryBuilder } from '@/lib/utils/qb';
import { getApexBets } from '@/actions/product/apex-bet/query';
import { ErrorMessage } from '@/components/status/error-message';
import { OpenBetsTable } from '@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table';
import { FormOutput } from '@/components/form/form-output';

// try some of these
export const revalidate = 0;
export const dynamic = 'force-dynamic';
export const fetchCache = 'force-no-store';
const uniqueKey = 'ob';

//searchParams are never updated

export default async function Page({ searchParams }: { searchParams: SearchParams }) {
  const query = queryBuilder(searchParams, uniqueKey);
  const { data, meta, error } = await getApexBets(query);
  return (
    <>
      searchParams:{JSON.stringify(searchParams)}
      <FormOutput title="Query Builder" values={query} />
      <PageHeading title="Open Bets" description="View all open bets" />
      <Container pb="xl">
        {error ? <ErrorMessage error={error} /> : null}
        <OpenBetsTable searchParams={searchParams} data={data} meta={meta} uniqueKey={uniqueKey} />
      </Container>
    </>
  );
}


```

**Comment by asmyshlyaev177** [comment]
Hi, also point out Next.js version, pages or App router?

**Comment by jimjamdev** [comment]
> @jimjamdev Hi, also point out Next.js version, pages or App router?

App Router / Nextjs 14.2.14

**Comment by jimjamdev** [comment]
Could be related to this part in the docs

```
export default async function Home({ searchParams }: { searchParams: object }) {
  return (
    <Form searchParams={searchParams} />
  )
}

// Form.tsx
'use client'
import React from 'react';
import { useUrlState } from 'state-in-url/next';
import { form } from './form';

const Form = ({ searchParams }: { searchParams: object }) => {
  const { urlState, setState, setUrl } = useUrlState(form, { searchParams });
}
```

**Comment by asmyshlyaev177** [comment]
Will be good to see ` <OpenBetsTable />` component as well, or even example on codesandbox (or any similar).

**Comment by jimjamdev** [comment]
I can't give to much as it's a company project. But let me see what i can give you.

`'use client';

import { Table } from '@mantine/core';
import { Meta, SearchParams } from '@/actions/global.schemas';
import { TableHeaderColumn } from '@/components/data-views/table-view/header-column';
import { FormatDate } from '@/components/format/format-date';
import { HeaderActions } from '@/components/data-views/header-actions';
import { ListSearch } from '@/components/data-views/search';
import { Pagination } from '@/components/data-views/pagination';
import { ApexBet } from '@/actions/product/apex-bet/schema';
import { TableEditButton } from '@/components/data-views/table-view/edit-button';
import { TableCreateButton } from '@/components/data-views/table-view/create-button';
import { TableActions, TableHeaderActions } from '@/components/data-views/table-view/table-actions';
import { OpenBetsCorrectButton } from '@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-correct-button';
import { BrokersListColumn } from '@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table/brokers-list-column';
import { OpenBetsStatusesFilter } from '@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table/statuses-filter';
import { BetStatusResultSelect } from '@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table/bet-status-select';
import { useClientState } from '@/lib/hooks/client-state';
import {
  openBetsInitialState,
  OpenBetsStateType,
} from '@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table/open-bets.state';
import { FormOutput } from '@/components/form/form-output';

export function OpenBetsTable({
  data,
  meta,
  uniqueKey = '',
  searchParams,
}: {
  data?: ApexBet[];
  meta: Meta;
  uniqueKey?: string;
  searchParams?: SearchParams;
}) {
  const headerColumns = [
    { label: 'Player', value: 'playerId' },
    { label: 'Date', value: 'eventDate' },
    { label: 'Broker', value: 'broker', sort: false },
    { label: 'Match', value: 'details.league' },
    { label: 'Stake', value: 'stake' },
    { label: 'Created', value: 'createdAt' },
    { label: 'Odds', value: 'price' },
    { label: 'Bet Status', value: 'status' },
    { label: 'Added By', value: 'addedBy' },
  ];

  const { urlState } = useClientState<OpenBetsStateType>(
    openBetsInitialState,
    uniqueKey,
    searchParams
  );

  return (
    <>
      <FormOutput title="Client state" values={urlState} />
      <HeaderActions
        left={<TableCreateButton />}
        right={
          <>
            <OpenBetsStatusesFilter searchParams={searchParams} uniqueKey={uniqueKey} />
            <ListSearch uniqueKey={uniqueKey} />
          </>
        }
      />

      <Table.ScrollContainer minWidth={800}>
        <Table highlightOnHover verticalSpacing="sm" stickyHeader>
          <Table.Thead>
            <Table.Tr>
              {headerColumns.map((column) => (
                <TableHeaderColumn key={column.value} uniqueKey={uniqueKey} column={column} />
              ))}
              <TableHeaderActions length={2} />
            </Table.Tr>
          </Table.Thead>
          <Table.Tbody>
            {data?.map((column) => (
              <Table.Tr key={column.id}>
                <Table.Td>{column.playerId}</Table.Td>
                <Table.Td>
                  <FormatDate date={column?.eventDate?.toString()} type="long" />
                </Table.Td>
                <Table.Td>
                  <BrokersListColumn name="Broker" brokers={column.brokers} />
                </Table.Td>
                <Table.Td>{column.bet}</Table.Td>
                <Table.Td>{column.stake}</Table.Td>
                <Table.Td>
                  <FormatDate date={column?.createdAt?.toString() as string} type="long" />
                </Table.Td>
                <Table.Td>{column.price}</Table.Td>
                <Table.Td>
                  <BetStatusResultSelect defaultValue={column.resultStatus} />
                </Table.Td>
                <Table.Td>{column.latestUpdatedBy}</Table.Td>
                <TableActions>
                  <TableEditButton id={column?.id?.toString() as string} />
                  <OpenBetsCorrectButton bet={column} />
                </TableActions>
              </Table.Tr>
            ))}
          </Table.Tbody>
        </Table>
      </Table.ScrollContainer>
      <Pagination uniqueKey={uniqueKey} total={meta?.totalCount} />
    </>
  );
}
`

this might be the issue as i wrap

```'use client';

import { useUrlState } from 'state-in-url/next';
import { useMemo } from 'react';
import { SearchParams } from '@/actions/global.schemas';

export function useClientState<T extends object>(
  initialState: T,
  uniqueKey: string,
  searchParams?: SearchParams
) {
  const { urlState, setUrl, setState } = useUrlState<Record<string, any>>(initialState, {
    searchParams,
    replace: true,
    scroll: false,
  });

  const setCustomUrl = (newState: Partial<T>) => {
    const prefixedState = Object.entries(newState).reduce(
      (acc, [key, value]) => ({
        ...acc,
        [`${uniqueKey}${key}`]: value,
      }),
      {} as Record<string, any>
    );

    setUrl(prefixedState);
  };

  // Cache the custom state using useMemo
  const customState = useMemo(() => {
    const state = {} as Record<string, any>;
    Object.entries(urlState).forEach(([key, value]) => {
      if (key.startsWith(uniqueKey)) {
        const unprefixedKey = key.slice(uniqueKey.length);
        state[unprefixedKey] = value;
      }
    });
    return state;
  }, [urlState, uniqueKey]);

  return {
    urlState: customState,
    setUrl: setCustomUrl,
    setState,
  };
}

**Comment by jimjamdev** [comment]
I'll test with the unwrapped state, and see what i get

**Comment by asmyshlyaev177** [comment]
I would check that state keys match, e.g. what defined in `'@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table/open-bets.state';` vs what going to `['${uniqueKey}${key}']: value,`.

Not sure, I would try to debug what happens if you reload the page, e.g. put console.log in some server component and see what it logs. After that with update from client component.

**Comment by asmyshlyaev177** [comment]
I will investigate it further, how exactly it works.

**Comment by jimjamdev** [comment]
> I would check that state keys match, e.g. what defined in `'@/app/(platform)/(modules)/apex-betting/(routes)/open-bets/components/open-bets-table/open-bets.state';` vs what going to `['${uniqueKey}${key}']: value,`. Intended that state defined at start as a const.
> 
> Not sure, I would try to debug what happens if you reload the page, e.g. put console.log in some server component and see what it logs. After that with update from client component.

I've removed the usage of that right now and just passing in the searchParams from the server i.e

``` const { urlState, setUrl } = useUrlState<OpenBetsStateType>(openBetsInitialState, searchParams);```

I'm trying to get a simplified version running on codesandbox, gimme 5

**Comment by asmyshlyaev177** [comment]
@jimjamdev I think I got it, by default library updates params only client side, because it's faster than whole client-to-server chain, try to opt out with `useUrlState(state, { useHistory: false })`

**Comment by jimjamdev** [comment]
> @jimjamdev I think I got it, by default library updates params only client side, because it's faster than whole client-to-server chain, try to opt out with `useUrlState(state, { useHistory: false })`
> 
> Probably need to update docs to point it out for Next.js. I'm starting to hate next.js sometimes.
> 
> If page behind auth check, e.g. not visible to search engines I would go heavily with client side things, easier and faster, but it's just my opinion.

Yeah, that was my first thought with a backoffice. But they we're specifically asking for the server caching. Mainly because their API is quite slow. Plus I'm going to be sharing all the mutation/actions/types across different projects, so kinda made sense to have a single solution to data.

**Comment by asmyshlyaev177** [comment]
Let me know if `useHistory` works.

**Comment by jimjamdev** [comment]
> Yeah, reusing common logic makes sense, whole flow with searchParams kind of cumbersome by default.
> 
> Let me know if `useHistory` works.

Didn't seem to make a difference

**Comment by jimjamdev** [comment]
Ok, maybe i need to dive in and do the update

**Comment by asmyshlyaev177** [comment]
> Ok, maybe i need to dive in and do the update

I did it few weeks back, process is half-automatic, nothing broken so far. Just minor changes. But depends on a codebase.
Some speed improvements for dev and for prod, turbopack still doesn't work correctly.

**Comment by jimjamdev** [comment]
> > Ok, maybe i need to dive in and do the update
> 
> I did it few weeks back, process is half-automatic, nothing broken so far. Just minor changes. But depends on a codebase. Some speed improvements for dev and for prod, turbopack still doesn't work correctly.

Are you using React 19?

**Comment by asmyshlyaev177** [comment]
Hm, actually doesn't work correctly with `useHistory: false` with v14, interesting.

**Comment by jimjamdev** [comment]
Hmmmmm, i need to see why. Time for a mass delete/cleanup

**Comment by asmyshlyaev177** [comment]
@jimjamdev let me know why, because it's just weird. Can check if searchParams get updated if do it manually - https://nextjs.org/docs/14/app/api-reference/functions/use-search-params#updating-searchparams .

**Comment by jimjamdev** [comment]
![image](https://github.com/user-attachments/assets/b05189b1-775b-4b5f-be11-5ad0f6fea947)

Top is printing out the current state that's then passed to the queryBuilder/fetch

Bottom is the current state when refreshed. I pass searchParams into the table and filter from the page

**Comment by asmyshlyaev177** [comment]
> I think i might have to create a single state and export it, then share that between components? The state is never being pre-filled with the current values

Should totally do it. Great if it works, also thinking about some elusive issues.

**Comment by jimjamdev** [comment]
Is that correct though?
```
const { urlState, setUrl } = useUrlState<OpenBetsStateType>(openBetsInitialState, {
    useHistory: false,
    searchParams,
    replace: true,
    scroll: false,
  });
  ````

from your knowledge it should be prefilling the state values from the searchParams?

**Comment by jimjamdev** [comment]
It could be an issue with my zod schema. This [] instead of a single key is a bit of a pita

```
someKey[eq] // lt, mt, gt, gte etc is causing me some issues 
```

**Comment by asmyshlyaev177** [comment]
> Is that correct though?
> 
> ```
> const { urlState, setUrl } = useUrlState<OpenBetsStateType>(openBetsInitialState, {
>     useHistory: false,
>     searchParams,
>     replace: true,
>     scroll: false,
>   });
> ```
> 
> from your knowledge it should be prefilling the state values from the searchParams?

Correct.

> It could be an issue with my zod schema. This [] instead of a single key is a bit of a pita

Can you elaborate what do you mean?

**Comment by jimjamdev** [comment]
Ok, it was my zod schema. To explain, to send data to the API it has operators, so a key: value could be like key[eq]: 1

So you have the initial state which i was using schema.parse i.e

```
import { ApexBetParams, ApexBetParamsSchema } from '@/actions/product/apex-bet/schema';

export type OpenBetsStateType = ApexBetParams;

export const openBetsInitialState: OpenBetsStateType = ApexBetParamsSchema.parse({});
```

However, changing it to below works:

```
// import { ApexBetParams, ApexBetParamsSchema } from '@/actions/product/apex-bet/schema';

export type OpenBetsStateType = {
  'statuses[eqm]': string[];
};

export const openBetsInitialState: OpenBetsStateType = {
  'statuses[eqm]': [],
};
```

So the issue now is trying to figure this out in zod. I might have to create multiple versions of the same key i.e statuses[eq], statuses[gt], etc

**Comment by jimjamdev** [comment]
Thanks for your help so far though. I'm thinking you've saved me weeks/months of work.

**Comment by asmyshlyaev177** [comment]
> Thanks for your help so far though. I'm thinking you've saved me weeks/months of work.

Glad to hear it.

Will publish a new version with some fixes, you can test it as beta if you like.

**Comment by jimjamdev** [comment]
Sure, what are the changes?

**Comment by asmyshlyaev177** [comment]
Trying to solve some elusive bug, related to how Next.js parse searchParams, and sometimes state "glitches", at least with `useHistory: true`, with Next.js router it's just generally slower.

**Comment by jimjamdev** [comment]
I don't know if this is any help or something you want to implement. But I'm wrapping your lib to give us functionality for a unique key.

```
// Will need updated
'use client';  

import { useUrlState } from 'state-in-url/next';
import { useMemo } from 'react';
import { SearchParams } from '@/actions/global.schemas';

export function useClientState<T extends object>(
  initialState: T,
  uniqueKey: string,
  searchParams?: SearchParams
) {
  const { urlState, setUrl, setState } = useUrlState<Record<string, any>>(initialState, {
    searchParams,
    useHistory: false,
    replace: true,
    scroll: false,
  });

  const setCustomUrl = (newState: Partial<T>) => {
    const prefixedState = Object.entries(newState).reduce(
      (acc, [key, value]) => ({
        ...acc,
        [`${uniqueKey}${key}`]: value,
      }),
      {} as Record<string, any>
    );

    setUrl(prefixedState);
  };

  // Cache the custom state using useMemo
  const customState = useMemo(() => {
    const state = {} as Record<string, any>;
    Object.entries(urlState).forEach(([key, value]) => {
      if (key.startsWith(uniqueKey)) {
        const unprefixedKey = key.slice(uniqueKey.length);
        state[unprefixedKey] = value;
      }
    });
    return state;
  }, [urlState, uniqueKey]);

  return {
    urlState: customState,
    setUrl: setCustomUrl,
    setState,
  };
}
```

Bascially you can have multiple tables on the same page, and a lot of the table have the same props to paginate/filter with, so the key just allows me to seperate the searchParams per table easily. The queryBuilder basically extracts all searchParams starting with the key

**Comment by jimjamdev** [comment]
> Trying to solve some elusive bug, related to how Next.js parse searchParams, and sometimes state "glitches", at least with `useHistory: true`, with Next.js router it's just generally slower.

Yeah, I noticed the speed impact as well. Nextjs certainly has it's issues. I think it's the re-render when the params change. 

There's some issues on their github e.g. https://github.com/vercel/next.js/issues/52870

**Comment by asmyshlyaev177** [comment]
Ah, so you need multiple states just with different keys, I see. I didn't expect this, but if it works, why not.

Next.js did make some not so good design decisions, hope they will refactor it later. Some issues with speed, in dev and in prod. Had to put this to config (maybe need to tune it)
```
  // only for dev mode
  onDemandEntries: {
    // period (in ms) where the server will keep pages in the buffer
    maxInactiveAge: 1 * 60 * 1000,
    // number of pages that should be kept simultaneously without being disposed
    pagesBufferLength: 6,
  },
```

**Comment by asmyshlyaev177** [comment]
@jimjamdev  Beta, just interesting if it's works well for you. - `npm install — save state-in-url@beta`

**Comment by jimjamdev** [comment]
Ok, will give it a test later tonight, balls deep right now

**Comment by asmyshlyaev177** [comment]
@jimjamdev or better to test 4.1.8 , seems like nothing broken, but feedback from more complex setup is better.
